src/Tasks/GetCompatiblePlatform.cs

```diff
@@ -0,0 +1,153 @@
+                }
+
+                HashSet<string> childPlatforms = new HashSet<string>();
+                foreach (string s in childPlatformOptions.Split(MSBuildConstants.SemicolonChar, StringSplitOptions.RemoveEmptyEntries))
+                {
+                    childPlatforms.Add(s);
+                }
+
+                string buildChildProjectAs = "";
+
+                // If the referenced project has a translation table, it came from the ProjectReference item's metadata.
+                // Prioritize that over the current project's translation table.
+                if (childPlatformLookupTable != null &&
+                        childPlatformLookupTable.ContainsKey(CurrentProjectPlatform) &&
+                        childPlatforms.Contains(childPlatformLookupTable[CurrentProjectPlatform]))
+                {
+                    buildChildProjectAs = childPlatformLookupTable[CurrentProjectPlatform];
+                    Log.LogMessage($"Found '{CurrentProjectPlatform}={buildChildProjectAs}' in the referenced project's translation table.");
+                }
+                // Current project's translation table follows
+                else if (translationTable != null &&
+                        translationTable.ContainsKey(CurrentProjectPlatform) &&
+                        childPlatforms.Contains(translationTable[CurrentProjectPlatform]))
+                {
+                    buildChildProjectAs = translationTable[CurrentProjectPlatform];
+                    Log.LogMessage($"Found '{CurrentProjectPlatform}={buildChildProjectAs}' in the current project's translation table.");
+                }
+                // Prefer matching platforms
+                else if (childPlatforms.Contains(CurrentProjectPlatform))
+                {
+                    buildChildProjectAs = CurrentProjectPlatform;
+                    Log.LogMessage($"Child and parent have the same platform.");
+                }
+                // AnyCPU if possible
+                else if (childPlatforms.Contains("AnyCPU"))
+                {
+                    buildChildProjectAs = "AnyCPU";
+                    Log.LogMessage($"Defaulting to AnyCPU.");
+                }
+                else
+                {
+                    // Keep NearestPlatform empty, log a warning. Common.CurrentVersion.targets will undefine 
+                    // Platform/PlatformTarget when this is the case.
+                    Log.LogWarningWithCodeFromResources("GetCompatiblePlatform.NoCompatiblePlatformFound", AssignedProjectsWithPlatform[i].ItemSpec);
+                }
+
+                AssignedProjectsWithPlatform[i].SetMetadata("NearestPlatform", buildChildProjectAs);
+                Log.LogMessage($"Project '{AssignedProjectsWithPlatform[i].ItemSpec}' will build with Platform: '{buildChildProjectAs}'");
+            }
+
+            return !Log.HasLoggedErrors;
+        }
+
+        private Dictionary<string, string> ExtractLookupTable(string stringTable)
```

Comment:
> Traditionally we try to be tolerant of whitespace in strings that might be defined in properties/item metadata. Can you do that here?

Do we typically trim the individual items in a semicolon-delimited list after we split it?

Reply:
> consider keeping a cache of string->IReadOnlyDictionary and checking whether we've looked at the stringTable before. That way we don't have to do these splits for every project.

Question on task lifetimes: Would global variables survive between task calls from the same project? Multiple projects? I worry about caching the tables statically, since Project A could define table X=Y and project B could define Z=A

Reply:
> Do we typically trim the individual items in a semicolon-delimited list after we split it?
> 
> 
> 
> Edit: Did it anyway. Don't see any harm to it.

👍🏻 this helps if you want to do something like

```xml
<P>
   Foo;
   Bar;
   $(P)
</P>
```

Reply:
> Question on task lifetimes: Would global variables survive between task calls from the same project? Multiple projects? I worry about caching the tables statically, since Project A could define table X=Y and project B could define Z=A

Great questions! Static variables live for the lifetime of the assembly load which is generally "the lifetime of the process" (you're not using an appdomain here and we don't use collectible ALCs on Core).

That might be ok and often is: if the data is small but recomputing it is expensive, spend the memory. Here recomputing is not super expensive so I dunno. On the flip side there aren't likely going to be like thousands of different input strings.

This can wait for sure.